### ISSUE TITLE
Can we use ILRepack for our dependencies instead of ExcelDnaPack?

### ISSUE THREAD

**Issue by wh1t3cAt1k** [Excel-DNA/ExcelDna#429]
Hi @govert !

This is a question more than a bug report, sorry if this is the wrong place to ask.

Currently we use ExcelDnaPack to pack everything into the XLL including our dependencies.

What we don't like about this process is the fact that we have to separately maintain the list of deps as `<Reference>`s in our .dna file (with `Pack="true"`).

I wonder if we could work around this by packing everything into our dll first using `ILRepack`, and then just referencing that as a single `<ExternalLibrary>` in the `<DnaLibrary>`.

**Comment by govert** [comment]
Hi @wh1t3cAt1k - I think it should work fine to make a single large assembly that you set as the `<ExternalLibrary>` The one downside I can see is that Excel-DNA scans through all the types in the `ExternalLibrary` assemblies to look for exports, ribbons etc. So if you have lots of extra types pulled in from the dependencies, your load time might increase a bit. You can certainly experiment with this.

What is the difficulty for you in setting up the <Reference ..> list? Do your dependencies change often?

In the new v1.6.0 previews we have done most of the work to support `<PackageReference>, so that you don't need a .dna file in your project anymore - it can be generated from defaults and properties in the .csproj file. I have yet to document this, but one bit there is an `<ExcelDnaReferences>` property where you can add a list of files for which `<Reference>` tags should be generated. I'm not sure whether we support wildcards there, but having some idea of how it would work best for your projects would be interesting.

**Comment by wh1t3cAt1k** [comment]
Hi @govert , our deps change several times a year, I would say.

My main rationale that it's another potential failure point - we exploded a couple of times when a build was working fine under debug, but in release mode (the only configuration where we enable packing) we discovered a little bit too late (sometimes "deeply at runtime") that something was forgotten in .dna.

So I would prefer to just have one central point of managing all the dependencies - and it seems that `ILRepack` takes care of merging in all your package references automagically.

Off-topic: regarding your comment about optional .dna files, where would `CustomUI` element go in our case in 1.6.0?

**Comment by govert** [comment]
With 1.6.0 you could still have a .dna file (then one is not generated from the .csproj properties). But I want to deprecate having `CustomUI` in the .dna file - I think it should rather go in code or as a resource in your own assembly.

**Comment by wh1t3cAt1k** [comment]
We don't mind moving it wherever, as long as it's supported in any way :)

**Comment by govert** [comment]
The one problem with packing "all .dll files in the output directory that are not ExternalLibrary files" is that there may be native or mixed assemblies which cannot be packed. Do you have such a problem, or do you always just pack every .dll in the output?

**Comment by garethhayter** [comment]
@Pavel I actually pack all my assemblies into a single DLL, then reference this in the .dna file. I do this as part of my automated build process. I don’t see any value adding this step to Excel DNA, seeing as I prefer fine-grained control over everything. What value do you see in Excel DNA automating this?

**Comment by wh1t3cAt1k** [comment]
> The one problem with packing "all .dll files in the output directory that are not ExternalLibrary files" is that there may be native or mixed assemblies which cannot be packed. Do you have such a problem, or do you always just pack every .dll in the output?

Yes, we have the sqlite native library which we don't pack for that purpose.

Therefore the mechanism in ExcelDNA should probably support some sort of blacklist.

**Comment by wh1t3cAt1k** [comment]
@garethhayter can I ask you what tool you're using to repack your assemblies?

I am having trouble with `ILRepack` because the assemblies stop being loadable, because they're merged in.

![image](https://user-images.githubusercontent.com/5159662/153598736-e51aa6dd-8a67-44c0-9794-3b0bda1e6b8a.png)

(Assembly.Load failure because that assembly is now part of the main .dll - worked fine with ExcelDnaPack)

@govert as I understand ExcelDNA pack does not merge assemblies' IL, it only embeds them?

**Comment by wh1t3cAt1k** [comment]
To update everyone, I encountered great success when I used [Costura.Fody](https://github.com/Fody/Costura) instead of ILRepack.

Apparently it does the same thing as ExcelDnaPack, but "automagically" includes all your "CopyLocal" deps, and allows you to manage a blacklist, too:

![image](https://user-images.githubusercontent.com/5159662/153704014-3bf151b1-b13b-44f2-9ba3-762e7dac4cc1.png)

I see no runtime issues and am happy with the result.
